### DISCUSSION TITLE
problems with network.WLAN.connect in v1.24.0 and 1.23.0

### DISCUSSION THREAD

**Discussion by ghost** [micropython/micropython#16089] [ESP32] [answered]
Hello, I'm just getting started with micropython on an ESP-WROOM-32 board; and I could not connect as a station to an AP when using 1.24.0 and 1.23.0. 

After creating an active wlan station, the AP I wanted to connect to would show up in scan() results, but attempts to connect(ssid, pass) never completed. the stations .status() was 211, and that didn't match any of the network  STAT_* constants. 

On a whim, I flash older firmware (1.19.1), and the esp32 connects to the AP as expected. v1.20.0 also works. 

Either I'm missing some configuration, or this a bug that I should report. If you have an esp32 as a wifi station with 1.24.0, could you please share the steps you took?

**Comment by davefes** [comment]
Using 1.23:
```py
#  06 Mar 2024 D. Festing


import network
import time
import config
import machine


sta_if = network.WLAN(network.STA_IF)


def connect():
    count = 0

# seems to help to start like this on ESP32
    sta_if.active(False)
    time.sleep(0.5)
    sta_if.active(True)
    time.sleep(0.5)

    if not sta_if.isconnected():
        print('connecting to hotspot...')
        time.sleep(0.1)

        try:
          # maybe should go after connect
#            sta_if.ifconfig((config.WiFi_device, '255.255.255.0', config.gateway, '8.8.8.8'))
            sta_if.connect(config.hotspot, config.password)
            sta_if.ifconfig((config.WiFi_device, '255.255.255.0', config.gateway, '8.8.8.8'))

        except OSError as error:
            try:
                with open('errors.txt', 'a') as outfile:
                    outfile.write(str(error) + '\n')
            except OSError:
                pass

        while (count < 20):
            count += 1

            if (sta_if.isconnected()):
                count = 0

                print(f'network config: {sta_if.ifconfig()}')

                break

            print('. ')
            time.sleep(1)


    if (count == 20):
        try:
            with open('errors.txt', 'a') as outfile:
                outfile.write('failed to connect after 20 tries' + '\n')
        except Exception as e:
            pass

        disconnect()  # or you could get errors
        print('machine reset')
        time.sleep(0.1)

        machine.reset()  # start from scratch


    rssi = sta_if.status('rssi')
    print(f'RSSI =  {rssi} dBm')
    time.sleep(0.1)

    if ((rssi < -95) or (rssi == 0)):
        print('signal level is below -95dBm')
        time.sleep(0.1)

        disconnect()  # or you could get errors
        print('machine reset')
        time.sleep(0.1)

        machine.reset()  # start from scratch


def disconnect():

    sta_if.disconnect()
    sta_if.active(False)
```

**Reply by ghost** [reply]
Thank you.  I will try the code and see what sort of errors (if any) get recorded.  I found the "# seems to help to start like this on ESP32" comment interesting. Turn it off and then turn it back on again.

**Reply by glenn20** [reply]
The reason for turning the wifi off then back on again is that after a micropython soft reset, the wifi hardware is not reset, and non-reproducible behaviour can happen if you try to bring up and configure a wifi device that is already configured. Setting the device inactive ensures you are starting from a known good state.

**Reply by ghost** [reply]
unfortunately, I got the same behavior as my last attempts with 1.23.  There was no wifi connection, and no error was recorded in the 'errors.txt' file. 

At this point, I'm going to say it is a bug that is affecting esp32-wroom-32 boards.

**Reply by robert-hh** [reply]
WiFi is working fine here, tested with several ESP32 board types. It is the same script which I use since MicroPython was ported to ESP32, similar to the example of the documentation:
```
WIFI_GATEWAY = "10.0.0.240"
GOOGLE_DNS = "8.8.8.8"
WIFI_SSID = "xxxx"
WIFI_PASSWD = "xxxxxxx"
WIFI_IP = "10.0.0.144"

def do_connect():
    import network
    import time

    wlan = network.WLAN(network.STA_IF) # create station interface
    wlan.active(True)       # activate the interface
    if not wlan.isconnected():      # check if the station is connected to an AP
        wlan.ifconfig((WIFI_IP, "255.255.255.0", WIFI_GATEWAY, GOOGLE_DNS))
        wlan.connect(WIFI_SSID, WIFI_PASSWD) # connect to an AP

        for _ in range(10):
            if wlan.isconnected():      # check if the station is connected to an AP
                break
            print('.', end='')
            time.sleep(1)
        else:
            print(" Connect attempt timed out\n")
            return
    print('\nnetwork config:', wlan.ifconfig())

do_connect()
```

**Reply by ghost** [reply]
@robert-hh can you verify that your code is working on v1.24.0?

**Reply by robert-hh** [reply]
It does work with 1.24.0, and I just compiled it with 1.25.0 preview 4, which has no changes to the ESP32 source C code and it works as well. It worked with 1.23. ....

**Reply by ghost** [reply]
it is still not working for me on 1.24.0. I replaced the line ` print('.', end='')` with `print( wlan.status())`, and as expected, the status code of 211 is the printed.  Something must have changed in the code base between 1.20.0 and 1.23.0 that is affecting my boards.  If I have some free time today, I will try to track  down the exact version release that introduced the bug affecting my boards.

**Reply by shariltumin** [reply]
These are the wlan.status() codes:
```
>>> network.STAT_ASSOC_FAIL
203
>>> network.STAT_BEACON_TIMEOUT
200
>>> network.STAT_CONNECTING
1001
>>> network.STAT_GOT_IP
1010
>>> network.STAT_HANDSHAKE_TIMEOUT
204
>>> network.STAT_IDLE
1000
>>> network.STAT_NO_AP_FOUND
201
>>> network.STAT_NO_AP_FOUND_IN_AUTHMODE_THRESHOLD
211
>>> network.STAT_NO_AP_FOUND_IN_RSSI_THRESHOLD
212
>>> network.STAT_NO_AP_FOUND_W_COMPATIBLE_SECURITY
210
>>> network.STAT_WRONG_PASSWORD
202
```

The 211 says "STAT_NO_AP_FOUND_IN_AUTHMODE_THRESHOLD" which I'm not sure what it really means. Try ```wlan.connect()``` with a wrong password, do you get 202?. Try also ```wlan.config('ssid')``` and ```wlan.config('channel')``` and compare them with the values from the ```wlan.scan()```. You can also try to set the ```network.country()``` to wlan country-code of your country, for example ```network.country('NO')```.

**Reply by ghost** [reply, marked as answer]
Sorry for the long delay in my response; I was busy working on my project using 1.20.0 and didn't have the wherewithal to debug the issue I was having with newer releases.

Happily, I found the issue and it was all thanks to you posting the undocumented 211 error constant of STAT_NO_AP_FOUND_IN_AUTHMODE_THRESHOLD.  My router was configured to use WPA for auth, and recent versions of esp-idf seem to only allow WPA2 and WPA3 connections. That is the 'threshold', and WPA is not within that threshold. 

I modified my router's config to use WPA2, and now I can use 1.24.1 without issue.

**Comment by FK-sauve** [comment]
config.gateway, '8.8.8.8' seems not good-》 this is Google dns not an ip gateway. Please see ipconfig (windows) or ifconfig on your pc settings to have the rigth settings.

**Reply by davefes** [reply]
When I first starting using USB dongles as a hotspot (HUAWEI-E8231) I often had poor connections to Gmail.  I was told that using the DNS server on the hotspot might be a problem,  I have had it working much better for the past year.  I could go back to using the DNS on the hotspot, but it takes weeks to find out if it is OK.

**Reply by glenn20** [reply]
I think @FK-sauve just got a bit confused with the semantics of the `ifconfig()` call. The 3rd argument is the gateway, which is stored in the `config.gateway` variable and the 4th argument is the DNS server.

**Comment by becorey** [comment]
I believe I ran into this same issue and it seemed to be related to memory. Connecting to wifi would work fine immediately after a fresh reboot, when gc.mem_free() showed about 80KB. But after importing a lot of modules, gc.mem_free() was down around 30KB, then I started getting weird problems connecting to wifi. This was on an ESP32-PICO-D4. If mem_free was really low like 10KB, it would report memory alloc errors, that helped me diagnose it. But in this "middle ground" of just not quite enough memory, I don't know what is failing internally but it wasn't reporting mem_alloc, it would report other things like Wifi Error 15, or 202, or 0x0101, or "internal error". 
I fixed it by doing various memory management techniques to free up more RAM. Refactored code, made my modules smaller, checked for allocating large variables, etc, and always doing gc.collect() right before sta_if.active(True) and sta_if.connect().
With a WROOM you can try a module with more PSRAM such as ESP32-S3-WROOM-1U-N4R2, R2 for 2MB PSRAM.

**Reply by davefes** [reply]
@becorey 

Interesting re the `gc.collect` statement.  In the applications that seem to just keep working I have `gc.collect`, within a few statements, before I call `connect()` in my posted code.  

Could be a good idea to put it in `def connect()` then I wouldn't have to remember.

**Reply by ghost** [reply]
the solution in my case was not memory related, but due to a change in the allowed network auth types https://github.com/orgs/micropython/discussions/16089#discussioncomment-11442609
